File: sdks/python/src/cvfile/embed/_chunk.py
"""Section-based markdown chunker — Python port of @cvfile/embed.

Chunk offsets and lengths are expressed as UTF-8 BYTE positions into the
markdown source (spec §5.1), not Unicode code-point indices. This matches the
JS chunker so a chunk emitted by one SDK and decoded by the other indexes back
into the same source bytes.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Literal

ChunkingMode = Literal["document", "section", "paragraph"]

_HEADING = re.compile(r"^(#{1,6})\s+(.+?)\s*$")


@dataclass(frozen=True, slots=True)
class MarkdownChunk:
    id: str
    text_offset: int
    text_length: int
    text: str


def chunk_markdown(markdown: str, *, mode: ChunkingMode = "section") -> list[MarkdownChunk]:
    data = markdown.encode("utf-8")
    if mode == "document":
        return [_document_chunk(data)]
    if mode == "paragraph":
        return _paragraph_chunks(data)
    return _section_chunks(data)


def _document_chunk(data: bytes) -> MarkdownChunk:
    return MarkdownChunk(id="document", text_offset=0, text_length=len(data), text=data.decode("utf-8"))


def _section_chunks(data: bytes) -> list[MarkdownChunk]:
    sections: list[MarkdownChunk] = []
    ids: set[str] = set()
    current_id: str | None = None
    current_start = 0
    cursor = 0

    def flush(end: int) -> None:
        nonlocal current_id
        if current_id is None:
            return
        text = data[current_start:end].decode("utf-8")
        if text.strip():
            sections.append(
                MarkdownChunk(
                    id=current_id,
                    text_offset=current_start,
                    text_length=end - current_start,
                    text=text,
                )
            )
        current_id = None

    for line in _split_with_offsets(data):
        match = _HEADING.match(line.text)
        if match:
            flush(line.offset)
            base = _slugify(match.group(2) or f"section-{len(sections) + 1}")
            current_id = _unique_id(base, ids)
            ids.add(current_id)
            current_start = line.offset
        elif current_id is None:
            current_id = _unique_id("preamble", ids)
            ids.add(current_id)
            current_start = line.offset
        cursor = line.offset + len(line.text.encode("utf-8"))
    flush(cursor)

    if not sections:
        return [_document_chunk(data)]
    return sections


def _paragraph_chunks(data: bytes) -> list[MarkdownChunk]:
    out: list[MarkdownChunk] = []
    ids: set[str] = set()
    cursor = 0
    n = 0
    while cursor < len(data):
        end = data.find(b"\n\n", cursor)
        if end == -1:
            end = len(data)
        text = data[cursor:end].decode("utf-8")
        if text.strip():
            base = _slugify(text.split("\n", 1)[0] or f"p-{n}")
            chunk_id = _unique_id(base, ids)
            ids.add(chunk_id)
            out.append(MarkdownChunk(id=chunk_id, text_offset=cursor, text_length=end - cursor, text=text))
            n += 1
        cursor = end + 2
    if not out:
        return [_document_chunk(data)]
    return out


@dataclass(frozen=True, slots=True)
class _Line:
    text: str
    offset: int


def _split_with_offsets(data: bytes) -> list[_Line]:
    """Split markdown into lines (keeping trailing newlines), tagged with each
    line's UTF-8 BYTE offset into the source."""
    out: list[_Line] = []
    offset = 0
    parts = data.split(b"\n")
    for i, part in enumerate(parts):
        with_nl = part + (b"\n" if i < len(parts) - 1 else b"")
        out.append(_Line(text=with_nl.decode("utf-8"), offset=offset))
        offset += len(with_nl)
    return out


_SLUG_RE = re.compile(r"[^a-z0-9]+")


def _slugify(s: str) -> str:
    base = _SLUG_RE.sub("-", s.lower()).strip("-")[:64]
    return base or "section"


def _unique_id(base: str, taken: set[str]) -> str:
    if base not in taken:
        return base
    n = 2
    while f"{base}-{n}" in taken:
        n += 1
    return f"{base}-{n}"

File: packages/embed-js/src/chunk.ts
/**
 * Section-based markdown chunker.
 *
 * Splits a markdown document on ATX headings (`#`, `##`, ...). Each chunk
 * carries the byte offset and length into the original UTF-8 source so a
 * downstream consumer can map a vector hit back to the exact substring
 * without re-tokenising. Pre-heading content becomes a "preamble" chunk.
 *
 * Per spec §5.1, `textOffset`/`textLength` are UTF-8 *byte* offsets into the
 * markdown source. We encode the document once with `TextEncoder`, track a
 * byte cursor while iterating lines (counting the trailing `\n` byte), and
 * derive each chunk's `text` by decoding the corresponding byte slice. This
 * keeps the offsets in agreement with the Go and Python SDKs for any
 * non-ASCII résumé.
 */

const encoder = new TextEncoder();
const decoder = new TextDecoder();

export type ChunkingMode = 'document' | 'section' | 'paragraph';

export interface MarkdownChunk {
  id: string;
  textOffset: number;
  textLength: number;
  text: string;
}

export interface ChunkOptions {
  mode?: ChunkingMode;
}

const HEADING = /^(#{1,6})\s+(.+?)\s*$/;

/** A source line plus its UTF-8 byte offset and byte length (including any trailing `\n`). */
interface ByteLine {
  text: string;
  offset: number;
  byteLength: number;
}

export function chunkMarkdown(markdown: string, opts: ChunkOptions = {}): MarkdownChunk[] {
  const mode = opts.mode ?? 'section';
  const bytes = encoder.encode(markdown);
  if (mode === 'document') {
    return [documentChunk(bytes)];
  }
  if (mode === 'paragraph') {
    return paragraphChunks(bytes);
  }
  return sectionChunks(bytes);
}

function documentChunk(bytes: Uint8Array): MarkdownChunk {
  return { id: 'document', textOffset: 0, textLength: bytes.byteLength, text: sliceText(bytes, 0, bytes.byteLength) };
}

function sectionChunks(bytes: Uint8Array): MarkdownChunk[] {
  const lines = splitWithByteOffsets(bytes);
  const sections: MarkdownChunk[] = [];
  let current: { id: string; start: number; end: number } | null = null;
  const ids = new Set<string>();

  function flush(end: number): void {
    if (!current) return;
    const text = sliceText(bytes, current.start, end);
    if (text.trim().length === 0) {
      current = null;
      return;
    }
    sections.push({ id: current.id, textOffset: current.start, textLength: end - current.start, text });
    current = null;
  }

  for (const line of lines) {
    const match = HEADING.exec(line.text);
    const lineEnd = line.offset + line.byteLength;
    if (match) {
      flush(line.offset);
      const id = uniqueId(slugify(match[2] ?? `section-${sections.length + 1}`), ids);
      ids.add(id);
      current = { id, start: line.offset, end: lineEnd };
      continue;
    }
    if (current === null) {
      const id = uniqueId('preamble', ids);
      ids.add(id);
      current = { id, start: line.offset, end: lineEnd };
    } else {
      current.end = lineEnd;
    }
  }
  flush(bytes.byteLength);

  if (sections.length === 0) {
    return [documentChunk(bytes)];
  }
  return sections;
}

function paragraphChunks(bytes: Uint8Array): MarkdownChunk[] {
  const out: MarkdownChunk[] = [];
  const ids = new Set<string>();
  const separator = encoder.encode('\n\n');
  let cursor = 0;
  let i = 0;
  while (cursor < bytes.byteLength) {
    let end = indexOfBytes(bytes, separator, cursor);
    if (end === -1) end = bytes.byteLength;
    const text = sliceText(bytes, cursor, end);
    if (text.trim().length > 0) {
      const id = uniqueId(slugify(text.split('\n')[0] ?? `p-${i}`), ids);
      ids.add(id);
      out.push({ id, textOffset: cursor, textLength: end - cursor, text });
      i += 1;
    }
    cursor = end + separator.byteLength;
  }
  if (out.length === 0) {
    return [documentChunk(bytes)];
  }
  return out;
}

/** Decode the UTF-8 byte slice `[start, end)` back into a string. */
function sliceText(bytes: Uint8Array, start: number, end: number): string {
  return decoder.decode(bytes.subarray(start, end));
}

/** Split UTF-8 bytes into lines, each tagged with its byte offset and byte length (newline included). */
function splitWithByteOffsets(bytes: Uint8Array): ByteLine[] {
  const newline = 0x0a; // '\n'
  const lines: ByteLine[] = [];
  let start = 0;
  for (let i = 0; i < bytes.byteLength; i += 1) {
    if (bytes[i] === newline) {
      const byteLength = i - start + 1;
      lines.push({ text: sliceText(bytes, start, i + 1), offset: start, byteLength });
      start = i + 1;
    }
  }
  if (start < bytes.byteLength) {
    lines.push({ text: sliceText(bytes, start, bytes.byteLength), offset: start, byteLength: bytes.byteLength - start });
  }
  return lines;
}

/** Find the byte index of `needle` in `haystack` at or after `from`, or -1. */
function indexOfBytes(haystack: Uint8Array, needle: Uint8Array, from: number): number {
  const last = haystack.byteLength - needle.byteLength;
  for (let i = from; i <= last; i += 1) {
    let matched = true;
    for (let j = 0; j < needle.byteLength; j += 1) {
      if (haystack[i + j] !== needle[j]) {
        matched = false;
        break;
      }
    }
    if (matched) return i;
  }
  return -1;
}

function slugify(s: string): string {
  return (
    s
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, '-')
      .replace(/^-+|-+$/g, '')
      .slice(0, 64) || 'section'
  );
}

function uniqueId(base: string, taken: Set<string>): string {
  if (!taken.has(base)) return base;
  let n = 2;
  while (taken.has(`${base}-${n}`)) n += 1;
  return `${base}-${n}`;
}
